tests/test_abstract.py

```diff
@@ -0,0 +1,103 @@
+    UnmeasurableVariable,
+    _get_measurable_outputs,
+    assign_custom_measurable_outputs,
+    noop_measurable_outputs_fn,
+)
+from aeppl.opt import construct_ir_fgraph
+
+
+def assert_equal_hash(classA, classB):
+    assert hash(classA) == hash(classA.id_obj)
+    assert hash(classB) == hash(classB.id_obj)
+    assert classA == classB
+    assert hash(classA) == hash(classB)
+
+
+@pytest.mark.parametrize(
+    "op, id_obj, class_dict",
+    [
+        (None, None, UnmeasurableVariable.__dict__),
+        (None, (1, 2), UnmeasurableVariable.__dict__),
+        (
+            NormalRV,
+            (NormalRV, noop_measurable_outputs_fn),
+            UnmeasurableVariable.__dict__,
+        ),
+    ],
+)
+def test_unmeasurable_variable_class(op, id_obj, class_dict):
+    A_dict = class_dict.copy()
+    B_dict = class_dict.copy()
+
+    if id_obj is not None:
+        A_dict["id_obj"] = id_obj
+        B_dict["id_obj"] = id_obj
+
+    if op is None:
+        parent_classes = (UnmeasurableVariable,)
+    else:
+        parent_classes = (op, UnmeasurableVariable)
+
+    A = type("A", parent_classes, A_dict)
+    B = type("B", parent_classes, B_dict)
+
+    assert_equal_hash(A, B)
+
+
+def test_unmeasurable_meta_hash_reassignment():
+    A_dict = UnmeasurableVariable.__dict__.copy()
+    B_dict = UnmeasurableVariable.__dict__.copy()
+
+    A_dict["id_obj"] = (1, 2)
+    B_dict["id_obj"] = (1, 3)
+
+    A = type("A", (UnmeasurableVariable,), A_dict)
+    B = type("B", (UnmeasurableVariable,), B_dict)
+
+    assert A != B
+    assert hash(A) != hash(B)
+
+    A.id_obj = (1, 3)
+
+    assert_equal_hash(A, B)
+
+
+def test_assign_custom_measurable_outputs():
+    srng = at.random.RandomStream(seed=2320)
+
+    X_rv = srng.normal(-10.0, 0.1, name="X")
+    Y_rv = srng.normal(10.0, 0.1, name="Y")
+
+    # manually checking assign_custom_measurable_outputs
+    unmeasurable_X = assign_custom_measurable_outputs(X_rv.owner).op
+    unmeasurable_Y = assign_custom_measurable_outputs(Y_rv.owner).op
+
+    assert_equal_hash(unmeasurable_X.__class__, unmeasurable_Y.__class__)
+    assert unmeasurable_X.__class__.__name__.startswith("Unmeasurable")
+    assert unmeasurable_X.__class__ in _get_measurable_outputs.registry
+
+    # checking assign_custom_measurable_outputs via compiled FunctionGraph
+
+    I_rv = srng.bernoulli(0.5, name="I")
+    Z_rv = at.stack([X_rv, Y_rv])[I_rv]  # mixture from stack
+
+    z_vv = Z_rv.clone()
+    i_vv = I_rv.clone()
+
+    fgraph, _, _ = construct_ir_fgraph({Z_rv: z_vv, I_rv: i_vv})
+
+    unmeasurable_X2 = fgraph.outputs[0].owner.inputs[2].owner.op
+    unmeasurable_Y2 = fgraph.outputs[0].owner.inputs[3].owner.op
+
+    assert_equal_hash(unmeasurable_X2.__class__, unmeasurable_Y2.__class__)
+    assert_equal_hash(unmeasurable_X.__class__, unmeasurable_X2.__class__)
+
+    with pytest.raises(ValueError):
+        assign_custom_measurable_outputs(
+            fgraph.outputs[0].owner.inputs[2].owner, lambda x: x
+        )
```

Comment:
Since this is testing mixture-specific functionality, it should go with the mixture tests; however, I don't think we need this kind of test in the first place, because the branches checked by this test should already be covered by the existing mixture tests.

Reply:
I see. I just forced push replacing this mixture test with one that verifies that passing an unmeasurable node into `assign_custom_measurable_outputs` does not do anything.